private/buf/cmd/buf/command/registry/registrylogin/registrylogin.go

```diff
@@ -211,11 +230,126 @@ func inner(
+			return "", errors.New("cannot perform an interactive login from a non-TTY device")
+		}
+		return "", err
+	}
+	return token, nil
+}
+
+// doBrowserLogin performs the device authorization grant flow via the browser.
+func doBrowserLogin(
+	ctx context.Context,
+	container appext.Container,
+	remote string,
+) (string, error) {
+	baseURL := "https://" + remote
+	clientName, err := getClientName()
+	if err != nil {
+		return "", err
+	}
+	externalConfig := bufapp.ExternalConfig{}
+	if err := appext.ReadConfig(container, &externalConfig); err != nil {
+		return "", err
+	}
+	appConfig, err := bufapp.NewConfig(container, externalConfig)
+	if err != nil {
+		return "", err
+	}
+	client := httpclient.NewClient(appConfig.TLS)
+	oauth2Client := oauth2.NewClient(baseURL, client)
+	// Register the device.
+	deviceRegistration, err := oauth2Client.RegisterDevice(ctx, &oauth2.DeviceRegistrationRequest{
+		ClientName: clientName,
+	})
+	if err != nil {
+		var oauth2Err *oauth2.Error
+		if errors.As(err, &oauth2Err) {
+			return "", fmt.Errorf("authorization failed: %s", oauth2Err.ErrorDescription)
+		}
+		return "", err
+	}
+	// Request a device authorization code.
+	deviceAuthorization, err := oauth2Client.AuthorizeDevice(ctx, &oauth2.DeviceAuthorizationRequest{
+		ClientID:     deviceRegistration.ClientID,
+		ClientSecret: deviceRegistration.ClientSecret,
+	})
+	if err != nil {
+		var oauth2Err *oauth2.Error
+		if errors.As(err, &oauth2Err) {
+			return "", fmt.Errorf("authorization failed: %s", oauth2Err.ErrorDescription)
+		}
+		return "", err
+	}
+	// Open the browser to the verification URI.
+	if err := browser.OpenURL(deviceAuthorization.VerificationURIComplete); err != nil {
+		return "", fmt.Errorf("failed to open browser: %w", err)
+	}
+	if _, err := fmt.Fprintf(
+		container.Stdout(),
+		`Opening your browser to complete authorization process.
+
+If your browser doesn't open automatically,
+please open this URL in a browser to complete the process:
+	%s
+`,
+		deviceAuthorization.VerificationURIComplete,
+	); err != nil {
+		return "", err
+	}
+	// Poll the token endpoint until the user has authorized the device.
+	deviceToken, err := oauth2Client.AccessDeviceToken(ctx, &oauth2.DeviceAccessTokenRequest{
+		ClientID:     deviceRegistration.ClientID,
+		ClientSecret: deviceRegistration.ClientSecret,
+		DeviceCode:   deviceAuthorization.DeviceCode,
+		GrantType:    oauth2.DeviceAuthorizationGrantType,
+	}, oauth2.AccessDeviceTokenWithPollingInterval(time.Duration(deviceAuthorization.Interval)*time.Second))
+	if err != nil {
+		var oauth2Err *oauth2.Error
+		if errors.As(err, &oauth2Err) {
+			return "", fmt.Errorf("authorization failed: %s", oauth2Err.ErrorDescription)
+		}
+		return "", err
+	}
+	return deviceToken.AccessToken, nil
+}
+
+// getClientName returns the client name for the device registration.
+func getClientName() (string, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", err
+	}
+	switch runtime.GOOS {
```

Comment:
nit: generally better pattern for this is using a variable and then having separate files for different OS's. See ie https://github.com/bufbuild/buf/tree/main/private/pkg/app app_unix.go app_windows.go